### PR TITLE
[boostrap-v5] Create types definition for Bootstrap 5

### LINKED ISSUE #45897: types for Bootstrap v5
Hello,

The alpha version of Bootstrap v5 was released more than a month ago and I would like to know if there are any types available for this version that remove Jquery.

If there are no types available yet, would it be possible to create some to be able to test the alpha version of v5 with TypeScript?

Have a nice day, 

Mael

@peterblazejewicz

### PR DESCRIPTION
**Pull request by martin-badin**
Closes: #45897

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test YOUR_PACKAGE_NAME`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

File: types/bootstrap/v5/modal.d.ts [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    enum Events {
+        /**
+         * This event fires immediately when the show instance method is called.
+         * If caused by a click, the clicked element is available as the
+         * relatedTarget property of the event.
+         */
+        show = 'show.bs.modal',
+
+        /**
+         * 	This event is fired when the modal has been made visible to the user
```
Comment (danger-public) [inline types/bootstrap/v5/modal.d.ts:45]: - :no_entry_sign: The root .editorconfig style specifies spaces for whitespace. Please use spaces in new or changed types.

**Comment by danger-public** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## bootstrap ([<kbd>unpkg</kbd>](https://unpkg.com/browse/bootstrap@latest/))
was missing the following properties:
1. Alert
1. Button
1. Carousel
1. Collapse
1. Dropdown

<details>
<summary>as well as these 7 other properties...</summary>
<p>Modal, Popover, Scrollspy, Tab, Toast, Tooltip, Util</p>
</details>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 74649da0e6d2c8c4e298ed74107373f8920b9979
</p>

**Comment by amcasey** [comment]
@peterblazejewicz Seems like you wanted to be tagged on PRs for these types.

**Comment by peterblazejewicz** [comment]
@martin-badin sorry for keeping you waiting, missed this for some reason (shame).
quick topics:
- what was the setup you've tried this?
- should types allow to use directly each component separately, as it ships as standalone UMD component itself?:

```ts
import Alert from 'bootstrap/js/dist/js/alert';

const alertElem = document.querySelector('.alert')!;
const alert = new Alert(alertElem);
alertElem.addEventListener('closed.bs.alert', () => {});
alert.close();
```
look for the 'swiper' types as example on complex ES6 setup we could use for reference,

**Comment by martin-badin** [comment]
@peterblazejewicz Thank you for your review.

- I installed types locally in my project but I have a problem that I don't know if it's possible to install just the version v5. What do you think?
- I added support to use each component separately. But I think you have the wrong path in your example. The path should be`bootstrap/js/dist/alert`

**Comment by martin-badin** [comment]
@peterblazejewicz I think would it be better if I move the old types for bootstrap 4 to the v4 folder and the newest types to the root of the package?

**Comment by peterblazejewicz** [comment]
> I think would it be better if I move the old types for bootstrap 4 to the v4 folder and the newest types to the root of the package?

Yes, I believe this what we discussed originally. That woudl require holding off released types until the non-beta/non-alfa version of v5 gets published to NPM. 
Yes, i believe v4 of BS is here to stay and be updated for a long time, so v4 is a must

**Comment by martin-badin** [comment]
I moved the old version of bootstrap to the v4 folder and the newest version of bootstrap to the root folder on this package.

File: types/bootstrap/js/dist/popover.d.ts [hunk 2]
```diff
@@ -0,0 +1,267 @@
+import * as Popper from 'popper.js';
+
+type ContentFunction = () => void;
```
Comment (pesoklp13) [inline types/bootstrap/js/dist/popover.d.ts:3]: Does it make sense to make three different types for the same signature?

File: types/bootstrap/js/dist/tooltip.d.ts [hunk 3]
```diff
@@ -0,0 +1,272 @@
+import * as Popper from 'popper.js';
+
+type PlacementFunction = () => void;
```
Comment (pesoklp13) [inline types/bootstrap/js/dist/tooltip.d.ts:3]: Same as for my previous comment. Do we need different types for the same signature?

File: types/bootstrap/js/dist/alert.d.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
+declare class Alert {
+    constructor(element: Element);
+
+    /**
+     * Closes an alert by removing it from the DOM. If the .fade and .show
+     * classes are present on the element, the alert will fade out before it
+     * is removed.
+     */
+    close: () => void;
```
Comment (rbuckton) [inline types/bootstrap/js/dist/alert.d.ts:9]: Are these supposed to be function-valued fields or methods? Making them function-valued fields affects subclassability and also affects assignability.

File: types/bootstrap/js/dist/alert.d.ts [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+         * completed.
+         */
+        closed = 'closed.bs.alert',
+    }
+
+    /**
+     * Static method which allows you to get the alert instance associated to a
+     * DOM element, you can use it like this: getInstance(alert)
+     */
+    function getInstance(element: Element): Alert;
```
Comment (rbuckton) [inline types/bootstrap/js/dist/alert.d.ts:35]: If this is a static method, why not declare it as a static method on the `Alert` class?

**Review by rbuckton — CHANGES_REQUESTED** [review]

File: types/bootstrap/index.d.ts [hunk 6]
```diff
@@ -1,654 +1,21 @@
-// Type definitions for Bootstrap 4.5
-// Project: https://github.com/twbs/bootstrap/, https://getbootstrap.com
-// Definitions by: denisname <https://github.com/denisname>
-//                 Piotr Błażejewicz <https://github.com/peterblazejewicz>
+// Type definitions for bootstrap 5.0
+// Project: https://v5.getbootstrap.com/
+// Definitions by: Martin Badin <https://github.com/martin-badin>
```
[the remark this answers is not in the document]
Comment (martin-badin) [reply types/bootstrap/index.d.ts:3]: He is not removed, only moved to the v4 folder. But I add him to this version too.

File: types/bootstrap/js/dist/popover.d.ts (hunk 2 above)
Comment (martin-badin) [reply types/bootstrap/js/dist/popover.d.ts:3]: I will remove it

File: types/bootstrap/js/dist/tooltip.d.ts (hunk 3 above)
Comment (martin-badin) [reply types/bootstrap/js/dist/tooltip.d.ts:3]: I will remove it

File: types/bootstrap/js/dist/alert.d.ts (hunk 5 above)
Comment (martin-badin) [reply types/bootstrap/js/dist/alert.d.ts:35]: good point. Thank you

File: types/bootstrap/js/dist/alert.d.ts (hunk 4 above)
Comment (martin-badin) [reply types/bootstrap/js/dist/alert.d.ts:9]: These properties are methods.

File: types/bootstrap/index.d.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (rbuckton) [reply types/bootstrap/index.d.ts]: Thanks. typescript-bot only picks up the maintainers from the root index.d.ts, IIRC.

File: types/bootstrap/js/dist/alert.d.ts (hunk 4 above)
Comment (rbuckton) [reply types/bootstrap/js/dist/alert.d.ts:9]: Then I would suggest writing them as this:
```suggestion
    close(): void;
```

**Review by rbuckton — CHANGES_REQUESTED** [review]

File: types/bootstrap/js/dist/alert.d.ts (hunk 4 above)
Comment (martin-badin) [reply types/bootstrap/js/dist/alert.d.ts]: Thank you, I updated PR

**Comment by martin-badin** [comment]
@rbuckton Could you please review again?

**Review by rbuckton — APPROVED** [review]

**Comment by martin-badin** [comment]
Ready to merge

**Pull request merged**